File: 2020/day05.py
import sys

#BFFFBBFRRR: row 70, column 7, seat ID 567.
#FFFBBBFRRR: row 14, column 7, seat ID 119.
#BBFFBBFRLL: row 102, column 4, seat ID 820.

test = (
'BFFFBBFRRR',
'FFFBBBFRRR',
'BBFFBBFRLL'
)

DEBUG = 'debug' in sys.argv

if 'test' in sys.argv:
    data = test
else:
    data = open('day05.txt').read().split('\n')[:-1]

def xlate(n):
    return int(n.replace('B','1').replace('F','0').replace('R','1').replace('L','0'),2)

#def xlate(n):
#    x = int(n[:7].replace('B','1').replace('F','0'),2)
#    y = int(n[7:].replace('R','1').replace('L','0'),2)
#    if DEBUG:
#        print( n, x, y, x*8+y )
#    return x*8+y

ids = [xlate(i) for i in data]
print( "Part 1:", max( ids ) )
ids.sort()
if DEBUG:
    print(ids)
last = 0
for i in ids:
    if last and i != last+1:
        print( "Part 2:", i-1 )
        break
    last = i

File: 2020/day05.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
using namespace std;

//BFFFBBFRRR: row 70, column 7, seat ID 567.
//FFFBBBFRRR: row 14, column 7, seat ID 119.
//BBFFBBFRLL: row 102, column 4, seat ID 820.

/*
test = (
'BFFFBBFRRR',
'FFFBBBFRRR',
'BBFFBBFRLL'
)
*/


int main()
{
    ifstream fin("day05.txt");
    string line;
    vector<int> seats;
    while( getline( fin, line ) )
    {
        int seat = 0;
        for( char c : line )
        {
            seat <<= 1;
            if( c == 'B' || c == 'R' )
                ++seat;
        }
        seats.push_back( seat );
#ifdef DEBUG
        cout << line << " " << seat << "\n";
#endif
    }
    sort( seats.begin(), seats.end() );
    cout << "Part 1: " << seats.back() << "\n";
    int last = 0;
    for( int seat : seats )
    {
        if( last && seat != last+1 )
            break;
        last = seat;
    }
    cout << "Part 2: " << (last+1) << "\n";
}
